src/neo/Ledger/Blockchain.cs

```diff
@@ -314,78 +324,27 @@ private VerifyResult OnNewBlock(Block block)
         {
             if (block.Index <= Height)
                 return VerifyResult.AlreadyExists;
-            if (block_cache.ContainsKey(block.Hash))
-                return VerifyResult.AlreadyExists;
-            if (block.Index - 1 >= header_index.Count)
+            if (block.Index - 1 > Height)
             {
                 AddUnverifiedBlockToCache(block);
                 return VerifyResult.UnableToVerify;
             }
-            if (block.Index == header_index.Count)
+            if (block.Index == Height + 1)
             {
                 if (!block.Verify(currentSnapshot))
                     return VerifyResult.Invalid;
-            }
-            else
-            {
-                if (!block.Hash.Equals(header_index[(int)block.Index]))
-                    return VerifyResult.Invalid;
-            }
-            block_cache.TryAdd(block.Hash, block);
-            if (block.Index == Height + 1)
-            {
-                Block block_persist = block;
-                List<Block> blocksToPersistList = new List<Block>();
-                while (true)
-                {
-                    blocksToPersistList.Add(block_persist);
-                    if (block_persist.Index + 1 >= header_index.Count) break;
-                    UInt256 hash = header_index[(int)block_persist.Index + 1];
-                    if (!block_cache.TryGetValue(hash, out block_persist)) break;
-                }
-
-                int blocksPersisted = 0;
-                foreach (Block blockToPersist in blocksToPersistList)
-                {
-                    block_cache_unverified.Remove(blockToPersist.Index);
-                    Persist(blockToPersist);
-
-                    // 15000 is the default among of seconds per block, while MilliSecondsPerBlock is the current
-                    uint extraBlocks = (15000 - MillisecondsPerBlock) / 1000;
-
-                    if (blocksPersisted++ < blocksToPersistList.Count - (2 + Math.Max(0, extraBlocks))) continue;
-                    // Empirically calibrated for relaying the most recent 2 blocks persisted with 15s network
-                    // Increase in the rate of 1 block per second in configurations with faster blocks
-
-                    if (blockToPersist.Index + 100 >= header_index.Count)
-                        system.LocalNode.Tell(new LocalNode.RelayDirectly { Inventory = blockToPersist });
-                }
+                block_cache.TryAdd(block.Hash, block);
```

Comment:
Here, `block_cache` will store the block to be persisted.

Reply:
But the block will be added soon.

Reply:
If we don't store here, this will cause our synchronization to slow down, as the other nodes must wait for it to persist. 
Or store the `header_index` here, too?

Reply:
Yes. We can update `header_index` at the beginning of `Persist()`.

Reply:
Have modified.